crates/subspace-service/src/sync_from_dsn/fast_sync.rs

```diff
@@ -0,0 +1,587 @@
+    ///     - download state for it,
+    /// - 4. add blocks with execution from the last segment,
+    ///     - notify Substrate sync by importing block with using BlockOrigin::NetworkBroadcast,
+    ///     - clear the block gap to prevent Substrate sync to download blocks from the start.
+    ///     - restart the Substrate sync with SyncMode::Full to enable it downloading full blocks.
+    ///  Note: retry for fast sync will degrade reputation if we have already announced blocks.
+    pub(crate) async fn sync(&self) -> Result<FastSyncResult<Block>, Error> {
+        debug!("Starting fast sync...");
+
+        // 1. Download the last two segments.
+        let mut reconstructor = Reconstructor::new().map_err(|error| error.to_string())?;
+        let segment_header_downloader = SegmentHeaderDownloader::new(self.node);
+
+        if let Err(error) = self
+            .download_segment_headers(&segment_header_downloader)
+            .await
+        {
+            let error = format!("Failed to download segment headers: {}", error);
+            return Err(Error::Other(error));
+        };
+
+        let Some(last_segment_index) = self.segment_headers_store.max_segment_index() else {
+            return Err(Error::Other("Can't get last segment index.".into()));
+        };
+
+        // Skip the fast sync if we lack the minimum required segment number
+        if last_segment_index <= SegmentIndex::ONE {
+            return Ok(FastSyncResult::skipped());
+        }
+
+        let last_segment_header = self
+            .segment_headers_store
+            .get_segment_header(last_segment_index)
+            .expect("Last segment index exists; qed");
+
+        let second_last_segment_index = last_segment_header.segment_index() - SegmentIndex::ONE;
+
+        let second_last_segment_blocks = download_and_reconstruct_blocks(
+            second_last_segment_index,
+            self.piece_getter,
+            &mut reconstructor,
+        )
+        .await?;
+
+        let blocks_in_second_last_segment = second_last_segment_blocks.len();
+
+        if blocks_in_second_last_segment < 2 {
+            return Err(Error::Other(
+                "Unexpected block array length for the second last segment.".into(),
+            ));
+        }
+
+        debug!(
+            "Second last segment blocks downloaded (SegmentId={}): {}-{}",
+            second_last_segment_index,
+            second_last_segment_blocks[0].0,
+            second_last_segment_blocks[blocks_in_second_last_segment - 1].0
+        );
+
+        let blocks = download_and_reconstruct_blocks(
+            last_segment_header.segment_index(),
+            self.piece_getter,
+            &mut reconstructor,
+        )
+        .await?;
+
+        let blocks_in_last_segment = blocks.len();
+
+        if blocks_in_last_segment < 2 {
+            return Err(Error::Other(
+                "Unexpected block array length for the last segment.".into(),
+            ));
+        }
```

Comment:
This check is unnecessary, the only thing we need is a single block. Other blocks can be imported even via regular Substrate sync without issues.

Here is the invariant from the spec you should check instead:
```suggestion
        if blocks.is_empty() {
            return Err(Error::Other(
                "Failed to reconstruct first block of the last segment".into(),
            ));
        }
```